roc/address.go

```diff
@@ -0,0 +1,73 @@
+package roc
+
+/*
+#cgo LDFLAGS: -lroc
+#include <roc/receiver.h>
+#include <roc/address.h>
+#include <roc/sender.h>
+#include <roc/log.h>
+#include <stdlib.h>
+*/
+import "C"
+
+import (
+	"runtime"
+	"unsafe"
+)
+
+func stringToCharArray(str string) []C.char {
+	charArray := make([]C.char, len(str))
+	for ind, r := range str {
+		charArray[ind] = (C.char)(r)
+	}
+	return charArray
+}
+
+// NewAddress parses the `ip`, `port` and `family` and initializes the Address object
+func NewAddress(family Family, ip string, port int) (*Address, error) {
+	a := new(Address)
+	a.mem = make([]byte, C.sizeof_roc_address)
+	a.raw = (*C.roc_address)(unsafe.Pointer(&a.mem))
+
+	cfamily := (C.roc_family)(family)
+	ip = safeString(ip)
+	cip := stringToCharArray(ip)
+	cport := (C.int)(port)
+	errCode := C.roc_address_init(a.raw, cfamily, (*C.char)(unsafe.Pointer(&cip[0])), cport)
+	runtime.KeepAlive(a.mem)
+
+	if errCode == 0 {
+		return a, nil
+	}
+	if errCode < 0 {
+		return nil, ErrInvalidArguments
+	}
+	return nil, ErrInvalidApi
+}
+
+func (a *Address) Family() (Family, error) {
+	f := C.roc_address_family(a.raw)
+	family := (Family)(f)
+	if family == AfInvalid {
+		return family, ErrInvalidArguments
+	}
+	return family, nil
+}
+
+func (a *Address) IP() (string, error) {
+	const buflen = 255
+	sIP := make([]byte, buflen)
+	res := C.roc_address_ip(a.raw, (*C.char)(unsafe.Pointer(&sIP)), buflen)
+	if res == nil {
+		return "", ErrInvalidArguments
+	}
+	return C.GoString(res), nil
```

Comment:
Go compiler has no way to know that `res` points to `sIP` data, so here we should insert `runtime.KeepAlive` manually:

```
if res == nil {
	return "", ErrInvalidArguments
}
ret := C.GoString(res)
runtime.KeepAlive(sIP) // ret is actually a pointer to sIP data
return ret, nil
```

Reply:
hmm, how does the test work then? I would like to have something failing before fixing, because otherwise it's unclear if my code fixes the issue or not :thinking:

Reply:
It works because the probability that GC is called *between* C.roc_address_ip and C.GoString, *and* that it will collect sIP during this invocation, is very low. This makes such bugs very unpleasant because they likely will reproduce rarely and only under specific load conditions.

And it's very hard to cover runtime.KeepAlive stuff with tests. Maybe only with intensive stress testing. Adding a stress test for roc-go in future is a good idea I think.

So our main tool for now is very careful code review.

I hope that memory-related bugs (like using `&mem` instead of `&mem[0]`) are easier to catch with tests.

Reply:
You can also try to run tests under race detector. We can even add it to the travis build.

Reply:
soo, here we shouldn't add keepAlive as well, following your arguments?
> unsafe.Pointer references are traceable by GC
    so *C.roc_address should be traceable as well
    cgo automatically inserts cgoUse() call for every pointer argument, which is superior to runtime.KeepAlive (in addition, it forces the pointer to escape to heap, to prevent it moving during stack reallocation)

Reply:
somehow this conversation gets hidden by github, that's why you probably aswered this question twice :thinking: 
I see your point :+1:

Reply:
Yeah, github duplicates a review comment if it's an answer to a hidden discussion. Unfortunately github still doesn't work very well with long discussions and review threads.